File: Roadmap/14 - FECHAS/typescript/roswer13.ts
// EJERCICIO:
// Crea dos variables utilizando los objetos fecha (date, o semejante) de tu lenguaje:
// - Una primera que represente la fecha (día, mes, año, hora, minuto, segundo) actual.
// - Una segunda que represente tu fecha de nacimiento (te puedes inventar la hora).
// Calcula cuántos años han transcurrido entre ambas fechas.

function getDates(): [Date, Date] {
    return [new Date(), new Date(1993, 1, 13, 6, 36, 34)];
}

function calcuatesYearsDifference(now: Date, birth: Date): number {
    var nowYear = now.getFullYear();
    var birthYear = birth.getFullYear();

    if (nowYear >= birthYear) {
        return nowYear - birthYear;
    }

    return birthYear - nowYear;
}

const [now, birth] = getDates();
console.log("Tiempo transcurrido desde el nacimiento: ", calcuatesYearsDifference(now, birth), " años.");


// DIFICULTAD EXTRA (opcional):
// Utilizando la fecha de tu cumpleaños, formatéala y muestra su resultado de
// 10 maneras diferentes. Por ejemplo:
// - Día, mes y año.
// - Hora, minuto y segundo.
// - Día de año.
// - Día de la semana.
// - Nombre del mes.

function dayIntToStringFormat(day: number): string {
    switch (day) {
        case 0: return "Sunday";
        case 1: return "Monday";
        case 2: return "Tuesday";
        case 3: return "Wednesday";
        case 4: return "Thursday";
        case 5: return "Friday";
        case 6: return "Saturday";
        default: return "Invalid day";
    }
}

function formatDates(date: Date) {
    return {
        day: date.getDate(),
        month: date.getMonth() + 1,
        year: date.getFullYear(),
        hours: date.getHours(),
        minutes: date.getMinutes(),
        seconds: date.getSeconds(),
        dayOfYear: Math.floor((date.getTime() - new Date(date.getFullYear(), 0, 0).getTime()) / 86_400_000),
        dayOfWeek: dayIntToStringFormat(date.getDay()),
        monthName: date.toLocaleString('en', { month: 'long' })
    };
}

const formattedDates = formatDates(birth);
console.log(`Dia, mes y año: ${formattedDates.day}/${formattedDates.month}/${formattedDates.year}`);
console.log(`Hora, minuto y segundo: ${formattedDates.hours}:${formattedDates.minutes}:${formattedDates.seconds}`);
console.log(`Dia del año: ${formattedDates.dayOfYear}`);
console.log(`Dia de la semana: ${formattedDates.dayOfWeek}`);
console.log(`Nombre del mes: ${formattedDates.monthName}`);

// Make testings for the functions.
function validateTets(bool: boolean, functionName?: string) {
    console.log(bool ? `✅ Test passed ${functionName}` : `❌ Test failed ${functionName}`);
}
function testGetDates() {
    // Arrange
    const [now, birth] = getDates();
    // Act
    // Assert
    validateTets(now instanceof Date, "testGetDates");
    validateTets(birth instanceof Date, "testGetDates");
}

function testCalcuatesYearsDifference() {
    // Arrange
    const now = new Date(2021, 1, 13, 6, 36, 34);
    const birth = new Date(1993, 1, 13, 6, 36, 34);
    // Act
    const result = calcuatesYearsDifference(now, birth);
    // Assert
    validateTets(result === 28, "testCalcuatesYearsDifference");
}

function testCalculateYearsDifferenceWhenNowIsBeforeBirth() {
    // Arrange
    const now = new Date(1993, 1, 13, 6, 36, 34);
    const birth = new Date(2021, 1, 13, 6, 36, 34);
    // Act
    const result = calcuatesYearsDifference(now, birth);
    // Assert
    validateTets(result === 28, "testCalculateYearsDifferenceWhenNowIsBeforeBirth");
}

function testCalculateDayIntToStringFormat() {
    // Arrange
    // Act
    const result = dayIntToStringFormat(0);
    // Assert
    validateTets(result === "Sunday", "testCalculateDayIntToStringFormat");
}

function testFormatDates() {
    // Arrange
    const date = new Date(1993, 1, 13, 6, 36, 34);
    // Act
    const result = formatDates(date);
    // Assert
    validateTets(result.day === 13, "testFormatDates");
    validateTets(result.month === 2, "testFormatDates");
    validateTets(result.year === 1993, "testFormatDates");
    validateTets(result.hours === 6, "testFormatDates");
    validateTets(result.minutes === 36, "testFormatDates");
    validateTets(result.seconds === 34, "testFormatDates");
    validateTets(result.dayOfYear === 44, "testFormatDates");
    validateTets(result.dayOfWeek === "Saturday", "testFormatDates");
    validateTets(result.monthName === "February", "testFormatDates");
}

function runTests() {
    console.log("Running tests...");
    testGetDates();
    testCalculateYearsDifferenceWhenNowIsBeforeBirth();
    testCalcuatesYearsDifference();
    testCalculateDayIntToStringFormat();
    testFormatDates();
    console.log("Tests finished.");
}

runTests();

File: Roadmap/14 - FECHAS/python/roswer13.py
# EJERCICIO:
# Crea dos variables utilizando los objetos fecha (date, o semejante) de tu lenguaje:
# - Una primera que represente la fecha (día, mes, año, hora, minuto, segundo) actual.
# - Una segunda que represente tu fecha de nacimiento (te puedes inventar la hora).
# Calcula cuántos años han transcurrido entre ambas fechas.
import unittest
from datetime import datetime

def get_dates() -> tuple:
    now = datetime.now()
    birth = datetime(1993, 2, 13, 6, 36, 34)
    return now, birth

def calcuates_years_difference(now: datetime, birth: datetime) -> int:
    nowYear = now.year
    birthYear = birth.year

    if nowYear>=birthYear:
        return nowYear - birthYear
    
    return birthYear - nowYear

if __name__ == '__main__':
    now, birth = get_dates()
    print(f"Tiempo transcurrido desde el nacimiento: {calcuates_years_difference(now, birth)} años.")

# DIFICULTAD EXTRA (opcional):
# Utilizando la fecha de tu cumpleaños, formatéala y muestra su resultado de
# 10 maneras diferentes. Por ejemplo:
# - Día, mes y año.
# - Hora, minuto y segundo.
# - Día de año.
# - Día de la semana.
# - Nombre del mes.
def calculate_date_format(birth: datetime, format: str) -> str:
    return birth.strftime(format)

if __name__ == '__main__':
    _, birth = get_dates()
    print(f"Día, mes y año: {calculate_date_format(birth, '%d/%m/%Y')}") # 13/02/1993
    print(f"Hora, minuto y segundo: {calculate_date_format(birth, '%H:%M:%S')}") # 06:36:34
    print(f"Día de año: {calculate_date_format(birth, '%j')}") # 044
    print(f"Día de la semana: {calculate_date_format(birth, '%A')}") # Saturday
    print(f"Nombre del mes: {calculate_date_format(birth, '%B')}") # February

# TESTS

class TestDate(unittest.TestCase):
    def test_get_dates(self):
        # Arrange
        # Act
        now, birth = get_dates()
        # Assert
        self.assertTrue(type(now) == datetime)
        self.assertTrue(type(birth) == datetime)

    def test_calculates_years_difference(self):
        # Arrange
        now = datetime(2021, 6, 30, 12, 0, 0)
        birth = datetime(1993, 6, 30, 8, 0, 0)
        # Act
        result = calcuates_years_difference(now, birth)
        # Assert
        self.assertTrue(result == 28)

    def test_calculate_date_format(self):
        # Arrange
        birth = datetime(1993, 2, 13, 6, 36, 34)
        # Act
        # Assert
        self.assertTrue(calculate_date_format(birth, '%d/%m/%Y') == "13/02/1993")
        self.assertTrue(calculate_date_format(birth, '%H:%M:%S') == "06:36:34")
        self.assertTrue(calculate_date_format(birth, '%j') == "044")
        self.assertTrue(calculate_date_format(birth, '%A') == "Saturday")
        self.assertTrue(calculate_date_format(birth, '%B') == "February")

if __name__ == '__main__':
    unittest.main()